distributed/worker.py

```diff
@@ -3499,30 +3516,16 @@ async def execute(self, key: str, *, stimulus_id: str) -> None:
                     result["traceback_text"],
                 )
 
-            self.transitions(recommendations, stimulus_id=stimulus_id)
-
             logger.debug("Send compute response to scheduler: %s, %s", ts.key, result)
 
-            if self.validate:
-                assert ts.state != "executing"
-                assert not ts.waiting_for_data
-
         except Exception as exc:
-            assert ts
             logger.error(
                 "Exception during execution of task %s.", ts.key, exc_info=True
             )
-            emsg = error_message(exc)
-            emsg.pop("status")
-            self.transition(
-                ts,
-                "error",
-                **emsg,
-                stimulus_id=stimulus_id,
-            )
-        finally:
-            self.ensure_computing()
-            self.ensure_communicating()
+            msg = error_message(exc)
+            recommendations = {ts: tuple(msg.values())}
+
+        return recommendations, []
```

Comment:
This differs from https://github.com/dask/distributed/issues/5895:

> - The `Worker.execute` method is modified such that it no longer performs any transition but instead returns appropriate `StateMachineEvents` that trigger the necessary handlers. For instance
>   - TaskFinished
>   - Rescheduled
>   - TaskErred

I could not find any benefit in implementing those events vs. just returning recommendations?

Reply:
The benefit is that we want to log the events and keep the recommendations as an internal detail of the to-be-defined `WorkerState` class. Not using recommendations here is one of the more important points of the design proposals

Reply:
Please have a look now